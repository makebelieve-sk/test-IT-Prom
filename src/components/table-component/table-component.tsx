import React, { useState, Dispatch } from 'react';
import Table from 'react-bootstrap/Table';

import { CellComponent } from '../cell-component/cell-component';
import { ElemType, DataType } from '../../types';


type TableComponentType = {
    data: DataType, 
    setData: Dispatch<DataType>
};

export const TableComponent: React.FC<TableComponentType> = ({ data, setData }) => {  
    const { header, body } = data;      
    
    const [editMode, setEditMode] = useState<string>(`hidden`);

    const handleClick = (elem: ElemType) => {
        setData({header: elem[0].subDep.header, body: elem[0].subDep.body});
    };

    const createString = () => {
        let newArr = new Array(header.length).fill(``);
        body.push(newArr);

        setData({header: header, body: body});
    };

    const deleteHandler = (e, el: string[]) => {
        if (e.button === 1) {
            e.preventDefault();

            let index = body.indexOf(el);
            body.splice(index, 1);
            
            setData({header: header, body: body});
        }
    };

    const handleSubmit = (e, el: string, i: number, j: number) => {
        e.preventDefault();

        setEditMode(`hidden`);

        body[i][j] = el;

        setData({header: header, body: body})
    };

    const editHandler = () => {
        setEditMode(`visible`);
    };

    if (header && body) {
        return (
            <div className="wrapper-table">
                <Table striped bordered hover className="table">
                    <thead>
                        <tr>
                            { header.map((el, i) => {
                                return <td className="table-td" key={`${el}-${i}`}>{ el }</td>
                            }) }
                        </tr>
                    </thead>
                    <tbody>
                        { body.map((el, i) => {
                            return (    
                                <tr 
                                    key={`${el}-${i}-1`} 
                                    onDoubleClick={editHandler}
                                    onMouseDown={(e) => deleteHandler(e, el)}                                    
                                >
                                    { el.map((elem, j) => {
                                        if (Array.isArray(elem)) {
                                            return (
                                                <CellComponent 
                                                    j={j}
                                                    i={i}
                                                    elem={elem[0].name}
                                                    handleSubmit={handleSubmit}
                                                    editMode={editMode}
                                                    key={`${elem}-${i}-${j}`}
                                                    clickProp={handleClick.bind(this, elem)}
                                                />
                                            )                                        
                                        } else {
                                            return (
                                                <CellComponent 
                                                    j={j}
                                                    i={i}
                                                    elem={elem}
                                                    handleSubmit={handleSubmit}
                                                    editMode={editMode}
                                                    key={`${elem}-${i}-${j}`}
                                                    clickProp={null}
                                                />
                                            );
                                        }
                                    }) }
                                </tr>
                            )                                                       
                        }) }
                        <tr><td onClick={createString.bind(this)} className="table-tr_create" colSpan={header.length}>Создать</td></tr>
                    </tbody>
                </Table>
            </div>
        );
    } else {
        return null;
    }
};