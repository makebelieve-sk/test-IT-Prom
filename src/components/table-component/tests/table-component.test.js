import React from 'react';
import ReactDOM from 'react-dom'; 
import renderer from 'react-test-renderer';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import { TableComponent } from '../table-component';

Enzyme.configure({ adapter: new Adapter() });

const mock = {
    data: {
        header: [`1`, `2`, `3`],
        body: [[`1`], [`2`]],
    },
    setData: jest.fn()
};

const setUp = () => shallow(<TableComponent 
    data={mock.data}
    setData={mock.setData}
/>);

it(`TableComponent correctly rendering with props`, () => {
    const tree = renderer  
        .create(<TableComponent 
            data={mock.data}
            setData={mock.setData}
        />)
        .toJSON();

    expect(tree).toMatchSnapshot();
});

describe('TableComponent correctly rendering and functioning', () => {
    let wrapper;
  
    beforeEach(() => {
        wrapper = setUp();
    });

    it(`+++ TableComponent renders without crashing`, () => {
        const div = document.createElement('div');
      
        ReactDOM.render(<TableComponent />, div);
        ReactDOM.unmountComponentAtNode(div);
    });

    it('+++ render the connected TableComponent', () => {
        expect(wrapper.length).toEqual(1);
    });
});