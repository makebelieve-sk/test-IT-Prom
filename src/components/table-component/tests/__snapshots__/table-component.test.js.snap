// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableComponent correctly rendering with props 1`] = `
<div
  className="table"
>
  <table
    className="table table-striped table-bordered table-hover"
  >
    <thead>
      <tr>
        <td>
          1
        </td>
        <td>
          2
        </td>
        <td>
          3
        </td>
      </tr>
    </thead>
    <tbody>
      <tr
        onDoubleClick={[Function]}
        onMouseDown={[Function]}
      >
        <td
          onClick={null}
        >
          1
          <form
            className="edit-mode-hidden"
            onSubmit={[Function]}
          >
            <input
              onChange={[Function]}
              type="text"
              value="1"
            />
          </form>
        </td>
      </tr>
      <tr
        onDoubleClick={[Function]}
        onMouseDown={[Function]}
      >
        <td
          onClick={null}
        >
          2
          <form
            className="edit-mode-hidden"
            onSubmit={[Function]}
          >
            <input
              onChange={[Function]}
              type="text"
              value="2"
            />
          </form>
        </td>
      </tr>
      <tr>
        <td
          onClick={[Function]}
        >
          Создать
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
